### PR TITLE
Update SmmPaging Protections with new tests.

### PR DESCRIPTION
**Pull request by joschock**
Add a new suite of tests to SMM paging tests to test various SMM protection capabilities.

### CODE REVIEW

File: UefiTestingPkg/UefiTestingPkg.dec [hunk 1]
```diff
@@ -13,15 +13,14 @@
 
 
 [Includes]
-  
+  Include
 
 [LibraryClasses]
- 
+
```
Comment (spbrogan) [inline UefiTestingPkg/UefiTestingPkg.dec:22]: missing the declaration

File: UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c [hunk 2]
```diff
@@ -11,14 +11,20 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
 #include <Library/SmmServicesTableLib.h>
 #include <Library/SmmMemLib.h>
-#include <Library/DebugLib.h>
-#include <Library/BaseMemoryLib.h>
+
+#include <Library/PlatformSmmProtectionsTestLib.h>
 
 #include <Protocol/SmmExceptionTestProtocol.h>
 
+#include <Register/Intel/ArchitecturalMsr.h>
```
Comment (spbrogan) [inline UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c]: do you need these?

Comment (joschock) [reply UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c]: Yes, used by the new test to validate that SmmEntryPoint is not writable.

File: UefiTestingPkg/UefiTestingPkg.dec (hunk 1 above)
Comment (joschock) [reply UefiTestingPkg/UefiTestingPkg.dec:22]: Fixed.

File: UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/App/SmmPagingProtectionsTestApp.c [hunk 3]
```diff
@@ -261,6 +288,322 @@ InvalidRangesShouldBeReadProtected (
...
+    //
+    PostReset = TRUE;
+    SetUsbBootNext ();
+    SaveFrameworkState (Framework, &PostReset, sizeof (PostReset));
+
+    // Grab the CommBuffer and fill it in for this test
+    Status = SmmMemoryProtectionsGetCommBuffer (&CommBuffer);
+    UT_ASSERT_NOT_EFI_ERROR (Status);
+
+    CommBuffer->Function = SMM_PROTECTIONS_READ_UNAUTHORIZED_IO;
```
Comment (corthon) [inline UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/App/SmmPagingProtectionsTestApp.c:366]: It seems like these "reset == success" functions could be rewritten as a single function that takes in the Function ID in the context. But maybe that would be more complicated later.

**Review by corthon — COMMENTED** [review]
Test looks good.

There's a LOT of layered changes here: code cleaning, new functions, refactors, etc. But at the end of the day... it's test code, so I'd be a bit more lenient on the size of the commit.

I'm with @spbrogan on waiting for @kuqin12 to report back and approving afterwards.

File: UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/App/SmmPagingProtectionsTestApp.c (hunk 3 above)
Comment (joschock) [reply UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/App/SmmPagingProtectionsTestApp.c:366]: Agreed. there's a lot of replicated "talk to SMM" boilerplate. But I didn't intend to rewrite the whole thing at the outset - just wanted to make myself satisfied that the new features worked as described. There is certainly room to rethink and refactor this whole thing in a vNext version.

File: UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c [hunk 4]
```diff
@@ -231,13 +257,245 @@ SmmMemoryProtectionsTestInvalidRange (
...
+{
+  UINT64 SmmBase;
+  VOID *SmmEntry;
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  SmmBase = AsmReadMsr64(MSR_IA32_SMBASE);
```
Comment (kuqin12) [inline UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c]: This is Intel centric. Might cause non page fault on other processors.

**Review by kuqin12 — APPROVED** [review]

File: UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c (hunk 4 above)
Comment (joschock) [reply UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c]: I will move it into the plat library.

Comment (joschock) [reply UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c]: Done; code moved to the plat library.

**Review by corthon — APPROVED** [review]

**Pull request merged**